core/dbt/task/retry.py

```diff
@@ -0,0 +1,113 @@
+from pathlib import Path
+
+from dbt.cli.flags import Flags
+from dbt.cli.types import Command as CliCommand
+from dbt.config import RuntimeConfig
+from dbt.contracts.results import NodeStatus
+from dbt.contracts.state import PreviousState
+from dbt.exceptions import DbtRuntimeError
+from dbt.graph import GraphQueue
+from dbt.task.base import ConfiguredTask
+from dbt.task.build import BuildTask
+from dbt.task.compile import CompileTask
+from dbt.task.generate import GenerateTask
+from dbt.task.run import RunTask
+from dbt.task.run_operation import RunOperationTask
+from dbt.task.seed import SeedTask
+from dbt.task.snapshot import SnapshotTask
+from dbt.task.test import TestTask
+
+RETRYABLE_STATUSES = {NodeStatus.Error, NodeStatus.Fail, NodeStatus.Skipped, NodeStatus.RuntimeErr}
+
+TASK_DICT = {
+    "build": BuildTask,
+    "compile": CompileTask,
+    "generate": GenerateTask,
+    "seed": SeedTask,
+    "snapshot": SnapshotTask,
+    "test": TestTask,
+    "run": RunTask,
+    "run-operation": RunOperationTask,
+}
+
+CMD_DICT = {
+    "build": CliCommand.BUILD,
+    "compile": CliCommand.COMPILE,
+    "generate": CliCommand.DOCS_GENERATE,
+    "seed": CliCommand.SEED,
+    "snapshot": CliCommand.SNAPSHOT,
+    "test": CliCommand.TEST,
+    "run": CliCommand.RUN,
+    "run-operation": CliCommand.RUN_OPERATION,
+}
+
+
+class RetryTask(ConfiguredTask):
+    def __init__(self, args, config, manifest):
+        super().__init__(args, config, manifest)
+
+        state_path = self.args.state or self.config.target_path
+
+        if self.args.warn_error:
+            RETRYABLE_STATUSES.add(NodeStatus.Warn)
```

Comment:
My understanding is:
- If the previous command passed `--warn-error`, then we will rerun tests or freshness checks that warned
- Otherwise, we will not rerun those warn-status tests/checks

I buy that - let's be sure to document it.